### PR TITLE
Add E0321 error explanation

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @Manishearth 
cc @AlisdairO 

Part of #24407

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -2677,6 +2677,28 @@ defined. For more information see the [opt-in builtin traits RFC](https://github
 .com/rust-lang/rfcs/blob/master/text/0019-opt-in-builtin-traits.md).
 "##,
 
+E0321: r##"
+A cross-crate trait with default impl wasn't implemented on a struct/enum
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: A cross-crate opt-out trait was implemented on something which wasn't a struct or enum

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -2677,6 +2677,28 @@ defined. For more information see the [opt-in builtin traits RFC](https://github
...
+struct NotSync;
+impl !Sync for NotSync {}
+
+unsafe impl Send for &'static NotSync {
+// error: cross-crate traits with a default impl, like `core::marker::Send`,
+//        can only be implemented for a struct/enum type, not
+//        `&'static NotSync`
+```
+
+Only structs and enums are permitted to impl Send and Sync, but not *all* of
```
Comment (Manishearth) [inline src/librustc_typeck/diagnostics.rs]: permitted to impl Send, Sync, and other opt-out trait, and the struct or enum must be local to the current crate. So, for example, `unsafe impl .... Rc ...` is not allowed.

**Comment by GuillaumeGomez** [comment]
Updated !

**Pull request merged**